### PR TITLE
[None][feat] Gemma4 MM: native vision + audio towers

### PR DESCRIPTION
**Pull request by Hudayday**
## Summary                                           
                                                                                                                                                                                                                    
  Per multimedia team's [PR](https://github.com/NVIDIA/TensorRT-LLM/pull/12932) feedback on the initial Gemma4 MM port, this PR replaces the temporary `AutoModel.from_config` HF fallback with **native TensorRT-LLM implementations** of the Gemma4 vision and audio
  (Conformer) towers, so that the multimodal path goes through TRT-LLM's own attention/MLP modules rather than re-entering HuggingFace at runtime.                                                                  
                                                    
  - **Native Gemma4 vision tower** (`modeling_gemma4_vision.py`) — replaces `AutoModel.from_config`; vision attention workspace now bound to the vision scale rather than inheriting the LM-side bound (avoids the  
  5589 GiB resize previously observed on 31B MMMU).         
  - **Native Gemma4 audio Conformer tower** (`modeling_gemma4_audio.py`) — replaces `AutoModel.from_config`; chunked local attention + rel-pos bias ported directly from the HF reference.
  - **HF Gemma4RMSNorm convention fix** (audio + vision) — Gemma4's MM-tower RMSNorm is the plain `w * x / rms`, **not** the Gemma3 LLM-side `(1 + w) * x / rms`. Wiring the correct norm convention is what brings 
  BLEU back into parity (E4B CoVoST2 1.90 → ~41.9 in earlier validation runs).
  - **`modeling_gemma4mm.py`** updated to import the native towers instead of the HF fallback.                                                                                                                      
  - **Test suite** `tests/unittest/_torch/multimodal/test_gemma4_multimodal.py` registered in `l0_b200.yml`, covering Tier 0/1/2 vision-tower parity plus the new RMSNorm regression tests.
  - Debug probes used during the port (`DEBUG_AUDIO_TOWER` env-gated stats) have been scrubbed before this commit; only Signed-off-by trailers in the history.                                                      
                                                                                                                                                                                                                    
  ## Alignment with HF tower baseline                       
                                                                                                                                                                                                                    
  The fix-mode (native towers) results have been **aligned with the HF-fallback baseline**. All completed fix-mode cells so far are within ±2pp of the HF baseline (well inside MMMU/CoVoST seed noise):                                                                                                                                         
                                                                                                                                                                                                                    
  | variant | task   | HF baseline | native fix | Δ      |                                                                                                                                                          
  |---------|--------|-------------|------------|--------|                                                                                                                                                          
  | E2B     | mmmu   | 42.67       | 42.11      | −0.56  |                                                                                                                                                          
  | E2B     | covost | 37.70 BLEU  | 37.78 BLEU | +0.08  |                                                                                                                                                          
  | E4B     | covost | 41.91 BLEU  | 41.91 BLEU | exact  |                                                                                                                                                          
  | 26B     | mmmu   | 58.33       | 58.56      | +0.23  |  

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added native audio and vision tower implementations for Gemma4 multimodal model, replacing previous placeholder approach with optimized TensorRT-LLM components.

* **Tests**
  * Expanded test coverage for multimodal audio/vision components, including forward pass validation, weight loading parity, and end-to-end generation testing.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14300?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_gemma4_audio.py [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+    attention_mask: Optional[torch.Tensor] = None
+
+
+# ---------------------------------------------------------------------------
+# RMSNorm: thin kwarg/dtype adapter over HF's ``Gemma4RMSNorm`` (see module
+# docstring for why this isn't the TRT-LLM ``RMSNorm`` module).
+# ---------------------------------------------------------------------------
+
+
+class _Gemma4AudioRMSNorm(_HFGemma4RMSNorm):
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4_audio.py]: Can you remove '_' from the class name please?

File: tensorrt_llm/_torch/models/modeling_gemma4_audio.py [hunk 2]
```diff
@@ -0,0 +1,713 @@
...
+from transformers.activations import ACT2FN
+from transformers.models.gemma4.modeling_gemma4 import Gemma4RMSNorm as _HFGemma4RMSNorm
+
+from ..model_config import ModelConfig
+from ..modules.linear import Linear
+from .modeling_utils import _load_weights_impl
+
+
+@dataclass
+class _AudioOutput:
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4_audio.py]: Also, for this one.

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py [hunk 3]
```diff
@@ -0,0 +1,819 @@
...
+        position_embeddings_btn = self.encoder.rotary_emb(inputs_embeds, pixel_position_ids)
+
+        # Flatten to (sum_valid, H) and drop padded tokens — encoder runs
+        # FULL attention with per-image cu_seqlens.
+        flat_embeds = inputs_embeds[valid]
+        cos_btn, sin_btn = position_embeddings_btn
+        flat_cos = cos_btn[valid]
+        flat_sin = sin_btn[valid]
+        flat_position_ids = pixel_position_ids[valid]
+        seq_lens_cpu = valid.sum(dim=-1).to(torch.int).cpu()
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Hi, this one can incure CPU-GPU sync. and CPU-GPU sync can hurt the performance. Is there any way we can do it other way?

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py [hunk 4]
```diff
@@ -0,0 +1,819 @@
...
+
+        # SigLip-style context-only metadata: kv_cache_manager=None, no decode
+        # phase. Re-prepared each forward with the actual per-image seq lens.
+        # Vision tower runs one image per call (mm forward loops per-image),
+        # so bounds are vision-scale, not LLM-scale. Using LLM max_num_tokens
+        # here makes the C++ attn workspace resize to TB-range and OOM
+        # (sized as max_num_requests * heads * max_num_tokens^2 * bytes).
+        self.metadata_cls = get_attention_backend(model_config.attn_backend).Metadata
+        self.attn_metadata = self.metadata_cls(
+            max_num_requests=1,
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Can I ask why you set 1?

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py [hunk 5]
```diff
@@ -0,0 +1,819 @@
...
+        )
+        self.pooler = Gemma4VisionPooler(self.config)
+        self.standardize = bool(getattr(self.config, "standardize", False))
+        if self.standardize:
+            self.register_buffer("std_bias", torch.zeros(self.config.hidden_size, dtype=dtype))
+            self.register_buffer("std_scale", torch.ones(self.config.hidden_size, dtype=dtype))
+
+        # SigLip-style context-only metadata: kv_cache_manager=None, no decode
+        # phase. Re-prepared each forward with the actual per-image seq lens.
+        # Vision tower runs one image per call (mm forward loops per-image),
```
Comment (yechank-nvidia) [inline tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Can I ask why is that? We should do cross-requests batching on images.

File: tensorrt_llm/_torch/models/modeling_gemma4_audio.py (hunk 1 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_audio.py]: Done. Renamed to Gemma4AudioRMSNorm.

File: tensorrt_llm/_torch/models/modeling_gemma4_audio.py (hunk 2 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_audio.py]: Done. Renamed to AudioOutput

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py (hunk 3 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Sync is API-mandated because it is using the TRTLLM-based Attention, which AttentionMetadata.prompt_lens is `Optional[List[int]] (host tensor)`, same in `modeling_siglip.py`, `modeling_qwen2vl.py`, `modeling_nemotron_nano.extract_feature_dynamic`. Even if we delayed it, `attn_metadata.prepare()` forces the materialisation.

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py (hunk 4 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Now scales with `model_config.max_num_requests` (like other mm models). The `=1` was a workaround for the unfused MHA path's O(B × T²) workspace; with the FMHA-on fix in the latest commit, the bump is safe.

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py (hunk 5 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: It was set that way for the same reason as the previous comment. The pattern now matches Gemma3vl / Qwen2VL / Qwen3VL / LlavaNext, supporting cross-request batching.

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py (hunk 3 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: I mean, yes in other modeling files, those are already on the cpu, so pinned_tensor to GPU would not occur CPU-GPU sync.

But it seems Gemma4 derives those values from GPU tensor to CPU which occurs CPU-GPU sync. and hurt performance very badly.

Is there anyway to derive those numbers from host tensors?

Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Thanks for the catch!

`Gemma4InputProcessor` now emits `image_seq_lens` as a CPU `List[int]` at preprocessing time (host-only, same shape as `Qwen2VL`'s `grid_thw.tolist()` path). The previous `valid.sum(-1).cpu()` GPU→CPU sync is gone; the tower now builds `seq_lens_cpu purely` from the host list and GPU is untouched.

E2E on B200 / 30-sample MMMU shows a small but measurable improvement

Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Thx for the update.

FYI, I just merged the claude skills for multimodal-models([https://github.com/NVIDIA/TensorRT-LLM/pull/13842](https://github.com/NVIDIA/TensorRT-LLM/pull/13842)). If you are okay, can you try?

**Review by yechank-nvidia — APPROVED** [review]
Thanks all for the hard work!

File: tensorrt_llm/_torch/models/modeling_gemma4_vision.py (hunk 3 above)
Comment (Hudayday) [reply tensorrt_llm/_torch/models/modeling_gemma4_vision.py]: Hi, I just used the agent skill to check this PR and fixed a few formatting and import mismatches. The agent check and CI should both be green now. Thanks for the suggestion!
